File: client_cpp/src/protocol.cpp
#include "protocol.hpp"
#include <cstring>

namespace proto {

// ==================== Big-endian encoding helpers ====================

static inline void putBE16(uint8_t* p, uint16_t v) {
    p[0] = uint8_t((v >> 8) & 0xFF);
    p[1] = uint8_t(v & 0xFF);
}

static inline void putBE32(uint8_t* p, uint32_t v) {
    p[0] = uint8_t((v >> 24) & 0xFF);
    p[1] = uint8_t((v >> 16) & 0xFF);
    p[2] = uint8_t((v >> 8) & 0xFF);
    p[3] = uint8_t(v & 0xFF);
}

static inline void putBE64(uint8_t* p, uint64_t v) {
    for (int i = 0; i < 8; i++) {
        p[i] = uint8_t((v >> (56 - 8 * i)) & 0xFF);
    }
}

// ==================== Big-endian decoding helpers ====================

static inline bool getBE16(const std::vector<uint8_t>& b, size_t& off, uint16_t& out) {
    if (off + 2 > b.size()) return false;
    out = (uint16_t(b[off]) << 8) | uint16_t(b[off + 1]);
    off += 2;
    return true;
}

static inline bool getBE32(const std::vector<uint8_t>& b, size_t& off, uint32_t& out) {
    if (off + 4 > b.size()) return false;
    out = (uint32_t(b[off]) << 24) | (uint32_t(b[off + 1]) << 16) | 
          (uint32_t(b[off + 2]) << 8) | uint32_t(b[off + 3]);
    off += 4;
    return true;
}

static inline bool getBE64(const std::vector<uint8_t>& b, size_t& off, uint64_t& out) {
    if (off + 8 > b.size()) return false;
    out = 0;
    for (int i = 0; i < 8; i++) {
        out = (out << 8) | uint64_t(b[off + i]);
    }
    off += 8;
    return true;
}

// ==================== Public encoding functions ====================

void putU16(std::vector<uint8_t>& b, uint16_t v) {
    size_t pos = b.size();
    b.resize(pos + 2);
    putBE16(b.data() + pos, v);
}

void putU32(std::vector<uint8_t>& b, uint32_t v) {
    size_t pos = b.size();
    b.resize(pos + 4);
    putBE32(b.data() + pos, v);
}

void putU64(std::vector<uint8_t>& b, uint64_t v) {
    size_t pos = b.size();
    b.resize(pos + 8);
    putBE64(b.data() + pos, v);
}

void putI32(std::vector<uint8_t>& b, int32_t v) {
    putU32(b, uint32_t(v));
}

void putDouble(std::vector<uint8_t>& b, double v) {
    uint64_t bits = 0;
    static_assert(sizeof(double) == 8, "double must be 8 bytes");
    std::memcpy(&bits, &v, 8);
    putU64(b, bits);
}

void putString(std::vector<uint8_t>& b, const std::string& s) {
    if (s.size() > 65535) return;
    putU16(b, uint16_t(s.size()));
    b.insert(b.end(), s.begin(), s.end());
}

void putPassword16(std::vector<uint8_t>& b, const std::string& s) {
    std::string t = s;
    if (t.size() > 16) t.resize(16);
    size_t pos = b.size();
    b.resize(pos + 16, 0);
    std::memcpy(b.data() + pos, t.data(), t.size());
}

// ==================== Public decoding functions ====================

bool getU16(const std::vector<uint8_t>& b, size_t& off, uint16_t& out) {
    return getBE16(b, off, out);
}

bool getU32(const std::vector<uint8_t>& b, size_t& off, uint32_t& out) {
    return getBE32(b, off, out);
}

bool getU64(const std::vector<uint8_t>& b, size_t& off, uint64_t& out) {
    return getBE64(b, off, out);
}

bool getI32(const std::vector<uint8_t>& b, size_t& off, int32_t& out) {
    uint32_t u;
    if (!getU32(b, off, u)) return false;
    out = int32_t(u);
    return true;
}

bool getDouble(const std::vector<uint8_t>& b, size_t& off, double& out) {
    uint64_t bits;
    if (!getU64(b, off, bits)) return false;
    std::memcpy(&out, &bits, 8);
    return true;
}

bool getString(const std::vector<uint8_t>& b, size_t& off, std::string& out) {
    uint16_t len;
    if (!getU16(b, off, len)) return false;
    if (off + len > b.size()) return false;
    out.assign(reinterpret_cast<const char*>(b.data() + off), len);
    off += len;
    return true;
}

bool getPassword16(const std::vector<uint8_t>& b, size_t& off, std::string& out) {
    if (off + 16 > b.size()) return false;
    const uint8_t* p = b.data() + off;
    size_t n = 16;
    while (n > 0 && p[n - 1] == 0) n--;
    out.assign(reinterpret_cast<const char*>(p), n);
    off += 16;
    return true;
}

// ==================== Message encoding/decoding ====================

std::vector<uint8_t> encode(const Message& m) {
    std::vector<uint8_t> out;
    out.reserve(HEADER_SIZE + m.body.size());

    putU32(out, m.h.magic);
    out.push_back(m.h.version);
    out.push_back(m.h.msgType);
    putU16(out, m.h.opCode);
    putU16(out, m.h.flags);
    putU16(out, m.h.status);
    putU64(out, m.h.requestId);
    putU32(out, m.h.bodyLen);

    out.insert(out.end(), m.body.begin(), m.body.end());
    return out;
}

bool decode(const std::vector<uint8_t>& raw, Message& out) {
    size_t off = 0;
    if (!getU32(raw, off, out.h.magic)) return false;
    if (out.h.magic != MAGIC) return false;
    if (off + 2 > raw.size()) return false;
    out.h.version = raw[off++];
    out.h.msgType = raw[off++];
    if (!getU16(raw, off, out.h.opCode)) return false;
    if (!getU16(raw, off, out.h.flags)) return false;
    if (!getU16(raw, off, out.h.status)) return false;
    if (!getU64(raw, off, out.h.requestId)) return false;
    if (!getU32(raw, off, out.h.bodyLen)) return false;
    if (off + out.h.bodyLen > raw.size()) return false;
    out.body.assign(raw.begin() + off, raw.begin() + off + out.h.bodyLen);
    return true;
}

// ==================== Utility functions ====================

std::string currencyToString(uint16_t c) {
    if (c == uint16_t(Currency::CNY)) return "CNY";
    if (c == uint16_t(Currency::SGD)) return "SGD";
    return "UNKNOWN";
}

std::string statusToString(uint16_t s) {
    switch (s) {
        case uint16_t(Status::OK): return "OK";
        case uint16_t(Status::ERR_BAD_REQUEST): return "Request format error (BAD_REQUEST)";
        case uint16_t(Status::ERR_AUTH): return "Authentication failed: name/account/password mismatch (AUTH)";
        case uint16_t(Status::ERR_NOT_FOUND): return "Account not found or already closed (NOT_FOUND)";
        case uint16_t(Status::ERR_CURRENCY): return "Currency mismatch (CURRENCY)";
        case uint16_t(Status::ERR_INSUFFICIENT_FUNDS): return "ERR_INSUFFICIENT_FUNDS";
        case uint16_t(Status::ERR_PASSWORD_FORMAT): return "Password format error: must be 1..16 bytes (PASSWORD_FORMAT)";
        default: return "Unknown error status=" + std::to_string(s);
    }
}

std::string opCodeToString(uint16_t op) {
    switch (op) {
        case uint16_t(OpCode::OPEN): return "OPEN";
        case uint16_t(OpCode::CLOSE): return "CLOSE";
        case uint16_t(OpCode::DEPOSIT): return "DEPOSIT";
        case uint16_t(OpCode::WITHDRAW): return "WITHDRAW";
        case uint16_t(OpCode::MONITOR_REGISTER): return "MONITOR_REGISTER";
        case uint16_t(OpCode::QUERY_BALANCE): return "QUERY_BALANCE";
        case uint16_t(OpCode::TRANSFER): return "TRANSFER";
        case uint16_t(OpCode::CALLBACK_UPDATE): return "CALLBACK_UPDATE";
        default: return "UNKNOWN_OP";
    }
}

} // namespace proto

File: server_java/src/Protocol.java
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

/**
 * Protocol definitions for the distributed banking system.
 * Uses big-endian (network byte order) for all multi-byte values.
 * 
 * Header format (24 bytes):
 *   magic:4 + version:1 + msgType:1 + opCode:2 + flags:2 + status:2 + requestId:8 + bodyLen:4
 */
public class Protocol {
    // Magic number: "BANK" in ASCII
    public static final int MAGIC = 0x42414E4B;
    public static final byte VERSION = 1;

    // Message types
    public static final byte MSG_REQUEST = 1;
    public static final byte MSG_REPLY = 2;
    public static final byte MSG_CALLBACK = 3;

    // Operation codes
    public static final short OP_OPEN = 1;              // Open account (non-idempotent)
    public static final short OP_CLOSE = 2;             // Close account (non-idempotent)
    public static final short OP_DEPOSIT = 3;           // Deposit (non-idempotent)
    public static final short OP_WITHDRAW = 4;          // Withdraw (non-idempotent)
    public static final short OP_MONITOR_REGISTER = 5;  // Monitor register
    public static final short OP_QUERY_BALANCE = 6;     // Query balance (idempotent)
    public static final short OP_TRANSFER = 7;          // Transfer (non-idempotent)
    public static final short OP_CALLBACK_UPDATE = 100; // Callback notification

    // Flags
    public static final short FLAG_AT_MOST_ONCE = 0x0001;

    // Status codes
    public static final short STATUS_OK = 0;
    public static final short STATUS_ERR_BAD_REQUEST = 1;
    public static final short STATUS_ERR_AUTH = 2;
    public static final short STATUS_ERR_NOT_FOUND = 3;
    public static final short STATUS_ERR_CURRENCY = 4;
    public static final short STATUS_ERR_INSUFFICIENT_FUNDS = 5;
    public static final short STATUS_ERR_PASSWORD_FORMAT = 6;

    // Currency types
    public static final short CUR_CNY = 0;
    public static final short CUR_SGD = 1;

    // Header size in bytes
    public static final int HEADER_SIZE = 24;

    /**
     * Message structure for request/reply/callback
     */
    public static class Message {
        public int magic;
        public byte version;
        public byte msgType;
        public short opCode;
        public short flags;
        public short status;
        public long requestId;
        public int bodyLen;
        public byte[] body;

        public Message() {
            this.magic = MAGIC;
            this.version = VERSION;
        }
    }

    /**
     * Encode a message to byte array for transmission
     */
    public static byte[] encode(Message m) {
        m.bodyLen = (m.body == null) ? 0 : m.body.length;
        ByteBuffer bb = ByteBuffer.allocate(HEADER_SIZE + m.bodyLen).order(ByteOrder.BIG_ENDIAN);
        bb.putInt(m.magic);
        bb.put(m.version);
        bb.put(m.msgType);
        bb.putShort(m.opCode);
        bb.putShort(m.flags);
        bb.putShort(m.status);
        bb.putLong(m.requestId);
        bb.putInt(m.bodyLen);
        if (m.bodyLen > 0) {
            bb.put(m.body);
        }
        return bb.array();
    }

    /**
     * Decode a byte array to message structure
     */
    public static Message decode(byte[] raw, int len) {
        if (len < HEADER_SIZE) return null;
        ByteBuffer bb = ByteBuffer.wrap(Arrays.copyOf(raw, len)).order(ByteOrder.BIG_ENDIAN);
        Message m = new Message();
        m.magic = bb.getInt();
        if (m.magic != MAGIC) return null;
        m.version = bb.get();
        m.msgType = bb.get();
        m.opCode = bb.getShort();
        m.flags = bb.getShort();
        m.status = bb.getShort();
        m.requestId = bb.getLong();
        m.bodyLen = bb.getInt();
        if (m.bodyLen < 0 || bb.remaining() < m.bodyLen) return null;
        m.body = new byte[m.bodyLen];
        bb.get(m.body);
        return m;
    }

    // ==================== Body encoding helpers ====================

    /**
     * Put a variable-length string (2-byte length prefix + UTF-8 bytes)
     */
    public static void putString(ByteBuffer bb, String s) {
        byte[] data = s.getBytes(StandardCharsets.UTF_8);
        if (data.length > 65535) throw new IllegalArgumentException("string too long");
        bb.putShort((short) data.length);
        bb.put(data);
    }

    /**
     * Put a fixed 16-byte password field (padded with zeros if shorter)
     */
    public static void putPassword16(ByteBuffer bb, String pw) {
        byte[] data = pw.getBytes(StandardCharsets.UTF_8);
        if (data.length == 0 || data.length > 16) {
            throw new IllegalArgumentException("password length must be 1..16 bytes");
        }
        byte[] fixed = new byte[16];
        System.arraycopy(data, 0, fixed, 0, data.length);
        bb.put(fixed);
    }

    // ==================== Body decoding helpers ====================

    /**
     * Get a variable-length string from buffer
     */
    public static String getString(ByteBuffer bb) {
        int len = Short.toUnsignedInt(bb.getShort());
        byte[] data = new byte[len];
        bb.get(data);
        return new String(data, StandardCharsets.UTF_8);
    }

    /**
     * Get a fixed 16-byte password field (trailing zeros trimmed)
     */
    public static String getPassword16(ByteBuffer bb) {
        byte[] fixed = new byte[16];
        bb.get(fixed);
        int n = 16;
        while (n > 0 && fixed[n - 1] == 0) n--;
        return new String(fixed, 0, n, StandardCharsets.UTF_8);
    }

    // ==================== Utility methods ====================

    public static String currencyToString(int c) {
        if (c == CUR_CNY) return "CNY";
        if (c == CUR_SGD) return "SGD";
        return "UNKNOWN";
    }

    public static String statusToString(int s) {
        switch (s) {
            case STATUS_OK: return "OK";
            case STATUS_ERR_BAD_REQUEST: return "Request format error (BAD_REQUEST)";
            case STATUS_ERR_AUTH: return "Authentication failed: name/account/password mismatch (AUTH)";
            case STATUS_ERR_NOT_FOUND: return "Account not found or already closed (NOT_FOUND)";
            case STATUS_ERR_CURRENCY: return "Currency mismatch (CURRENCY)";
            case STATUS_ERR_INSUFFICIENT_FUNDS: return "ERR_INSUFFICIENT_FUNDS";
            case STATUS_ERR_PASSWORD_FORMAT: return "Password format error: must be 1..16 bytes (PASSWORD_FORMAT)";
            default: return "Unknown error status=" + s;
        }
    }

    public static String opCodeToString(int op) {
        switch (op) {
            case OP_OPEN: return "OPEN";
            case OP_CLOSE: return "CLOSE";
            case OP_DEPOSIT: return "DEPOSIT";
            case OP_WITHDRAW: return "WITHDRAW";
            case OP_MONITOR_REGISTER: return "MONITOR_REGISTER";
            case OP_QUERY_BALANCE: return "QUERY_BALANCE";
            case OP_TRANSFER: return "TRANSFER";
            case OP_CALLBACK_UPDATE: return "CALLBACK_UPDATE";
            default: return "UNKNOWN_OP(" + op + ")";
        }
    }
}
